add methods to get or create categories and tags

kariae opened the pull request:
## Why are we adding these methods?

These methods (`get_or_create_category` and `get_or_create_tag`) provide a robust way to handle taxonomy term creation during migrations, with several key benefits:

1. **Idempotency**: Both methods ensure that duplicate terms aren't created, making them safe to run multiple times during migrations
2. **Validation**: They include proper validation for required fields and handle error cases gracefully
3. **Consistency**: They provide a consistent interface for creating both categories and tags, making the migration code more maintainable
4. **Error Handling**: They return proper WP_Error objects when something goes wrong, making it easier to handle errors in migration scripts
5. **Unique Identification**: They support unique identifiers for both categories and tags, allowing for reliable term tracking and deduplication during migrations

## Changes

- Added `get_or_create_category` and `get_or_create_tag` methods to the Taxonomy class
- Added support for unique identifiers in both methods to enable reliable term tracking
- Added comprehensive test suite for both methods, including unique identifier functionality

## Testing Instructions

### Manual Testing

#### Categories

1. **Creating a new category with a unique identifier**:
```php
$taxonomy = new \Newspack\MigrationTools\Logic\Taxonomy();
$cat_id = $taxonomy->get_or_create_category([
    'cat_name' => 'Test Category',
    'category_description' => 'Test Description',
    'category_nicename' => 'test-category'
], 'unique-cat-id-123');
// Should return a new category ID
```

2. **Getting an existing category by unique identifier**:
```php
$cat_id = $taxonomy->get_or_create_category([
    'cat_name' => 'Different Name' // Name doesn't matter when using unique identifier
], 'unique-cat-id-123');
// Should return the same ID as above, preserving the original name
```

3. **Creating a new category without a unique identifier**:
```php
$cat_id = $taxonomy->get_or_create_category([
    'cat_name' => 'Test Category',
    'category_description' => 'Test Description',
    'category_nicename' => 'test-category'
]);
// Should return a new category ID
```

4. **Getting an existing category by name**:
```php
$cat_id = $taxonomy->get_or_create_category([
    'cat_name' => 'Test Category'
]);
// Should return the same ID as above
```

#### Tags

1. **Creating a new tag with a unique identifier**:
```php
$taxonomy = new \Newspack\MigrationTools\Logic\Taxonomy();
$tag_id = $taxonomy->get_or_create_tag([
    'name' => 'Test Tag',
    'description' => 'Test Description',
    'slug' => 'test-tag'
], 'unique-tag-id-123');
// Should return a new tag ID
```

2. **Getting an existing tag by unique identifier**:
```php
$tag_id = $taxonomy->get_or_create_tag([
    'name' => 'Different Name' // Name doesn't matter when using a unique identifier
], 'unique-tag-id-123');
// Should return the same ID as above, preserving the original name
```

3. **Creating a new tag without a unique identifier**:
```php
$tag_id = $taxonomy->get_or_create_tag([
    'name' => 'Test Tag',
    'description' => 'Test Description',
    'slug' => 'test-tag'
]);
// Should return a new tag ID
```

4. **Getting an existing tag by name**:
```php
$tag_id = $taxonomy->get_or_create_tag([
    'name' => 'Test Tag'
]);
// Should return the same ID as above
```

#### Error Cases

1. **Missing required fields**:
```php
$result = $taxonomy->get_or_create_category([]);
// Should return WP_Error with 'missing_cat_name' code

$result = $taxonomy->get_or_create_tag([]);
// Should return WP_Error with 'missing_tag_name' code
```

2. **Duplicate unique identifiers**:
```php
// First call creates the term
$result = $taxonomy->get_or_create_tag([
    'name' => 'Test Tag'
], 'unique-tag-id-123');

// Second call with the same unique identifier returns the same term
$result2 = $taxonomy->get_or_create_tag([
    'name' => 'Test Tag'
], 'unique-tag-id-123');
// Should return the same ID as the first call

// Third call with the same name but a different unique identifier
$result3 = $taxonomy->get_or_create_tag([
    'name' => 'Test Tag'
], 'unique-tag-id-456');
// Should return WP_Error with 'term_exists' code
```

3. **Invalid parent category**:
```php
$result = $taxonomy->get_or_create_category([
    'cat_name' => 'Test Category',
    'category_parent' => 999999 // Non-existent category ID
]);
// Should return WP_Error with 'wrong_parent_category_term_id' code
```

### Automated Testing

The test suite can be run using PHPUnit:

```bash
composer test
```

The tests cover:
- Creating tags with just a name
- Creating tags with all optional fields
- Getting existing tags (idempotency)
- Error cases (missing name, empty name)
- Handling duplicate slugs
- Unique identifier functionality:
  - Creating terms with unique identifiers
  - Retrieving terms by unique identifier
  - Handling terms with the same name but different unique identifiers
  - The optional nature of unique identifiers
  - Proper storage and retrieval of unique identifiers in the term meta

## Checklist

- [x] Code follows WordPress coding standards
- [x] Added comprehensive test coverage
- [x] Added proper documentation
- [x] Tested manually
- [x] All tests pass
- [x] Unique identifier functionality tested and documented

kariae committed:
feat(Taxonomy): add methods to get or create categories and tags with error handling

kariae committed:
feat(Taxonomy): implement unique identifier support for categories and tags

- Added unique identifier parameters to `get_or_create_category` and `get_or_create_tag` methods.
- Introduced constants for unique category and tag identifier meta keys.
- Implemented a new method `get_term_id_by_unique_identifier` to retrieve terms by their unique identifiers.
- Enhanced tests to cover scenarios involving unique identifiers for both categories and tags.

iuravic commented on src/Logic/Taxonomy.php:
```diff
@@ -214,6 +226,148 @@ public function get_or_create_category_by_name_and_parent_id( string $cat_name,
+	/**
+	 * Gets or creates a category by its name and parent term_id.
+	 *
+	 * The array is the same as wp_insert_category accepts. https://developer.wordpress.org/reference/functions/wp_insert_category
```
A tiny nitpick ;) could we please put this description directly in the @param description, i.e.
```
* @param array  $data              The data to create the category with. The array is the same as wp_insert_category 
*                                  accepts. https://developer.wordpress.org/reference/functions/wp_insert_category.
```

And same thing for the other method `get_or_create_tag` 🙏

iuravic commented on src/Logic/Taxonomy.php:
```diff
@@ -214,6 +226,148 @@ public function get_or_create_category_by_name_and_parent_id( string $cat_name,
+	 * Note that you *have* to provide at least the 'cat_name' field.
+	 *
+	 * @param array  $data              The data to create the category with.
+	 * @param string $unique_identifier A unique identifier for your category – can be any string, but should be unique.
```
Not sure why our PHPCS isn't picking this up, but could we update `@param string $unique_identifier` to contain possible null value, i.e. `@param ?string $unique_identifier`?

And same thing for the other method `get_or_create_tag`, please.

iuravic reviewed (changes requested):
This looks really good and really nice! 👍 

I left two tiny details about docblock descriptions, which can be cleared and updated in 5 seconds ;)

And I have a question -- do we need the separate `TaxonomyMeta` class? Seems to me that perhaps `get_term_id_from_key_and_value` is still purely `Taxonomy` stuff. Do we foresee more things specific to taxonomy Meta coming up in the future, or could we merge this class into `Taxonomy`?

Thanks -- please feel free to merge after updating the tiny param comments, and thank you! 🙏

kariae committed:
refactor(Taxonomy): consolidate taxonomy meta handling and remove TaxonomyMeta class

- Replaced usage of TaxonomyMeta with a new private method in Taxonomy for retrieving term IDs by meta key and value.
- Updated method documentation to clarify parameters and usage.
- Removed the TaxonomyMeta class as its functionality has been integrated into Taxonomy.

kariae commented:
> do we need the separate TaxonomyMeta class? 

Good point. No, I don't forecast any usage for taxonomy metas more than that. I removed that class and did the doc fixes.

iuravic reviewed (approved):
🚀 ! 👏 !
